tools/cldr-apps/src/main/java/org/unicode/cldr/web/SubtypeToURLMap.java

```diff
@@ -0,0 +1,396 @@
+        try (InputStream fis = new FileInputStream(fn);
+            InputStreamReader inputStreamReader = new InputStreamReader(fis, Charset.forName("UTF-8"));
+            BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
+            ) {
+            return SubtypeToURLMap.getInstance(bufferedReader);
+        }
+    }
+
+    /**
+     * Fetch from a URL.
+     *
+     * @param resource
+     * @return
+     * @throws IOException
+     */
+    public static SubtypeToURLMap getInstance(URL resource) throws IOException, URISyntaxException {
+        if(resource.toString().endsWith(".txt")) {
+            // plain text
+            Class<?> classes[] = {InputStream.class};
+            try (InputStream is  = (InputStream)resource.getContent(classes);
+                Reader isr = new InputStreamReader(is, StandardCharsets.UTF_8);
+                BufferedReader br = new BufferedReader(isr);
+                ) {
+                return new SubtypeToURLMap(br);
+            }
+        } else if(resource.getProtocol().equals("http") || resource.getProtocol().equals("https")) {
+
+            Document doc = Jsoup.connect(resource.toString()).get();
+            SubtypeToURLMap newMap = getInstance(resource, doc);
+            try(PrintWriter pw = new PrintWriter(new File("/tmp/mymap.txt"), "UTF-8")) { // TODO: in later Java, we can use StandardCharsets.UTF8
+                newMap.write(pw);
+            }
+            return newMap;
+        } else {
+            // assume HTML to parse
+            Class<?> classes[] = {InputStream.class};
+            try (InputStream is  = (InputStream)resource.getContent(classes);
+                ) {
+                Document doc = Jsoup.parse(is, "UTF-8", resource.toString());
+                return getInstance(resource, doc);
+            }
+        }
+    }
+
+    private static SubtypeToURLMap getInstance(URL resource, Document doc) throws IOException {
+        StringBuffer sb = new StringBuffer();
+        doc.select("div code").forEach(n ->
+            n.textNodes()
+                .forEach(tn -> sb.append(tn.text()).append('\n')));
+        System.err.println("Read " + sb.length() + " chars from " + resource.toString());
+        try (Reader sr = new StringReader(sb.toString());
+            BufferedReader br = new BufferedReader(sr);) {
+            return new SubtypeToURLMap(br);
+        }
+    }
+    static final String DEFAULT_URL = /*CLDRConfig.getInstance().getProperty("CLDR_SUBTYPE_URL", */
+        "https://sites.google.com/site/cldr/development/subtypes" /* ) */;
+
+    private static String CACHE_SUBTYPE_FILE = "urlmap-cache.txt";
+
+    private final static class SubtypeToURLMapHelper {
+        static File cacheFile = new File(CLDRConfig.getInstance().get("CLDRHOME"), CACHE_SUBTYPE_FILE);
+        static SubtypeToURLMap INSTANCE = make(); // not final, may be reloaded.
+        static SubtypeToURLMap make() {
+            SubtypeToURLMap map = null;
```

Comment:
here again, subroutines would be nice to avoid having to scroll up and down to see the whole method; the subroutine names could help to clarify what it's doing from a top-down perspective

Reply:
makes sense.

Reply:
refactored